In [44]:
import pandas as pd
import numpy as np

## Reading files

In [84]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [85]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [86]:
wh = pd.concat([wh2018,wh2019])

In [87]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [88]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [89]:
result = pd.DataFrame()

In [90]:
for year_ in wh.loc[ : , "year" ].drop_duplicates():
    wh_year = wh.loc[ wh['year'] == year_ , : ]    
    for column_ in wh_year.columns:        
        percentage =  np.sum(pd.isna( wh_year.loc[ : , column_ ]  )) / wh_year.shape[0]
        dfrow = pd.DataFrame({
            'year': [ year_ ]
            , 'column' : [ column_ ]
            , 'percentage.of.nans' : [ percentage ]
        })
        result = pd.concat([result, dfrow])

In [91]:
result

year                        column  percentage.of.nans
0  2018                  Overall rank            0.000000
0  2018             Country or region            0.000000
0  2018                         Score            0.032051
0  2018                GDP per capita            0.057692
0  2018                Social support            0.096154
0  2018       Healthy life expectancy            0.076923
0  2018  Freedom to make life choices            0.076923
0  2018                    Generosity            0.108974
0  2018     Perceptions of corruption            0.044872
0  2018                          year            0.000000
0  2019                  Overall rank            0.000000
0  2019             Country or region            0.000000
0  2019                         Score            0.089744
0  2019                GDP per capita            0.019231
0  2019                Social support            0.006410
0  2019       Healthy life expectancy            0.089744
0  2019  Freedom to make life choices            0.070513
0  2019                    Generosity            0.025641
0  2019     Perceptions of corruption            0.038462
0  2019                          year            0.000000

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [92]:
for year_ in wh.loc[ : , "year" ].drop_duplicates():
    wh_year = wh.loc[ wh['year'] == year_ , : ]
    
    print( "YEAR: {0}".format(year_) )
    
    print( wh_year.describe() )
    

YEAR: 2018
       Overall rank       Score  GDP per capita  Social support  \
count    156.000000  151.000000      147.000000      141.000000   
mean      78.500000    5.354060        0.892748        1.205915   
std       45.177428    1.109532        0.399153        0.306070   
min        1.000000    2.905000        0.000000        0.000000   
25%       39.750000    4.451500        0.598500        1.047000   
50%       78.500000    5.358000        0.940000        1.258000   
75%      117.250000    6.154000        1.214500        1.462000   
max      156.000000    7.632000        2.096000        1.592000   

       Healthy life expectancy  Freedom to make life choices  Generosity  \
count               144.000000                    144.000000  139.000000   
mean                  0.597576                      0.456514    0.181360   
std                   0.250700                      0.162312    0.099147   
min                   0.000000                      0.000000    0.000000   
25%  

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

In [93]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

In [94]:
columns_to_use = ['Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption']

In [95]:
wh_limpio  = pd.DataFrame()
for year_ in wh.loc[ : , "year" ].drop_duplicates():
    wh_year = wh.loc[ wh['year'] == year_ , : ].copy()
    print( "Working with: {0}".format(year_) )
    for c_ in columns_to_use:        
        mean_column = np.mean( wh_year.loc[ : , c_ ] )
        print( "Column: {0} has an mean of {1}".format(c_,mean_column) )        
        wh_year.loc[ np.isnan( wh_year[c_] ) , c_ ] = mean_column
    
    wh_limpio = pd.concat([ wh_limpio.copy() , wh_year.copy() ])

Working with: 2018
Column: Score has an mean of 5.354059602649008
Column: GDP per capita has an mean of 0.8927482993197281
Column: Social support has an mean of 1.2059148936170214
Column: Healthy life expectancy has an mean of 0.5975763888888888
Column: Freedom to make life choices has an mean of 0.45651388888888866
Column: Generosity has an mean of 0.1813597122302159
Column: Perceptions of corruption has an mean of 0.11285906040268458
Working with: 2019
Column: Score has an mean of 5.434028169014085
Column: GDP per capita has an mean of 0.9007581699346409
Column: Social support has an mean of 1.208264516129032
Column: Healthy life expectancy has an mean of 0.7229647887323947
Column: Freedom to make life choices has an mean of 0.39544137931034484
Column: Generosity has an mean of 0.18509210526315784
Column: Perceptions of corruption has an mean of 0.10905333333333334


In [106]:
wh_limpio.loc[ np.isnan( wh_limpio['Score'] ) , : ]

Empty DataFrame
Columns: [Overall rank, Country or region, Score, GDP per capita, Social support, Healthy life expectancy, Freedom to make life choices, Generosity, Perceptions of corruption, year]
Index: []

In [107]:
wh.loc[ np.isnan( wh['Score'] ) , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
5               6       Netherlands    NaN           1.361           1.488   
7               8       New Zealand    NaN           1.268             NaN   
52             53            Latvia    NaN           1.148           1.454   
89             90            Jordan    NaN             NaN           1.265   
128           129           Armenia    NaN           0.816           0.990   
11             12        Costa Rica    NaN           1.034           1.441   
12             13            Israel    NaN           1.276           1.455   
42             43          Colombia    NaN           0.985           1.410   
53             54       South Korea    NaN           1.301           1.219   
66             67          Pakistan    NaN           0.677           0.886   
71             72             Libya    NaN           1.044           1.303   
100           101            Jordan    NaN           0.837           1.225   
104           105              Laos    NaN           0.764           1.030   
113           114             Niger    NaN           0.138           0.774   
127           128              Mali    NaN           0.385           1.105   
135           136            Uganda    NaN           0.332           1.069   
137           138            Zambia    NaN           0.578           1.058   
140           141           Liberia    NaN           0.073           0.922   
150           151             Yemen    NaN           0.287           1.163   

     Healthy life expectancy  Freedom to make life choices  Generosity  \
5                      0.878                         0.638       0.333   
7                      0.876                         0.669       0.365   
52                     0.671                         0.363       0.092   
89                     0.645                         0.468       0.130   
128                    0.666                         0.260       0.077   
11                     0.963                         0.558       0.144   
12                     1.029                         0.371       0.261   
42                     0.841                         0.470       0.099   
53                     1.036                         0.159       0.175   
66                     0.535                         0.313       0.220   
71                     0.673                         0.416       0.133   
100                    0.815                         0.383       0.110   
104                    0.551                         0.547       0.266   
113                    0.366                         0.318       0.188   
127                    0.308                         0.327       0.153   
135                    0.443                         0.356       0.252   
137                    0.426                         0.431       0.247   
140                    0.443                           NaN       0.233   
150                    0.463                         0.143       0.108   

     Perceptions of corruption  year  
5                        0.295  2018  
7                        0.389  2018  
52                       0.066  2018  
89                       0.134  2018  
128                      0.028  2018  
11                       0.093  2019  
12                       0.082  2019  
42                       0.034  2019  
53                       0.056  2019  
66                       0.098  2019  
71                       0.152  2019  
100                      0.130  2019  
104                      0.164  2019  
113                      0.102  2019  
127                      0.052  2019  
135                      0.060  2019  
137                        NaN  2019  
140                      0.033  2019  
150                      0.077  2019

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [62]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro

In [105]:
pd.DataFrame( {"x":[1,2],"y":[3,4]} ) - pd.DataFrame( {"x":[10,16],"y":[77,18]} )

x   y
0  -9 -74
1 -14 -14

In [172]:
x = wh_limpio.sort_values(["Country or region","year"] , ascending = False )
x["rank"] = x.groupby(["Country or region"])["year"].rank( method = 'first' , ascending = False )
x = x.loc[ x["rank"] == 1.0 , : ].reset_index(drop=True)

In [173]:
y = wh_limpio.sort_values(["Country or region","year"] , ascending = True )
y["rank"] = y.groupby(["Country or region"])["year"].rank( method = 'first' , ascending = True )
y = y.loc[ y["rank"] == 1.0 , : ].sort_values(["Country or region"] , ascending = False ).reset_index(drop=True)

In [174]:
x.index = x['Country or region']
y.index = x['Country or region']

In [175]:
x = x.drop(['Country or region','year','rank'] ,axis = 1)
y = y.drop(['Country or region','year','rank'] ,axis = 1)

In [176]:
x - y

Overall rank     Score  GDP per capita  Social support  \
Country or region                                                           
Zimbabwe                      2 -0.029000           0.009           0.020   
Zambia                       13  1.057028           0.016           0.011   
Yemen                        -1  2.079028          -0.155           0.090   
Vietnam                      -1  0.072000           0.026          -0.019   
Venezuela                     6 -0.099000          -0.036          -0.042   
...                         ...       ...             ...             ...   
Argentina                    18 -0.302000           0.019          -0.036   
Angola                        0  0.000000           0.000           0.000   
Algeria                       4 -0.084000           0.023           0.006   
Albania                      -5  0.133000           0.031           0.031   
Afghanistan                   9 -0.429000           0.018          -0.020   

                   Healthy life expectancy  Freedom to make life choices  \
Country or region                                                          
Zimbabwe                          0.474965                        -0.045   
Zambia                            0.131000                        -0.072   
Yemen                             0.120000                        -0.101   
Vietnam                           0.149000                        -0.075   
Venezuela                         0.148000                         0.021   
...                                    ...                           ...   
Argentina                         0.137000                        -0.099   
Angola                            0.000000                         0.000   
Algeria                           0.098000                         0.009   
Albania                           0.084000                        -0.036   
Afghanistan                       0.106000                        -0.085   

                   Generosity  Perceptions of corruption  
Country or region                                         
Zimbabwe                0.019                  -0.010000  
Zambia                  0.026                   0.027053  
Yemen                   0.025                  -0.035859  
Vietnam                -0.030                  -0.006000  
Venezuela               0.008                  -0.005000  
...                       ...                        ...  
Argentina               0.004                  -0.062859  
Angola                  0.000                   0.000000  
Algeria                 0.018                  -0.021000  
Albania                 0.029                  -0.005000  
Afghanistan            -0.033                  -0.011000  

[160 rows x 8 columns]